text/0000-consensus-api.md

```diff
@@ -0,0 +1,320 @@
+simplify their presentation and explanation. Each method is implemented with a
+pair of (Request, Response) messages and the Consensus Proxy is responsible for
+handling this transformation within the Validator.
+
+The methods presented below are only meant to describe roughly what the
+high-level interactions between the Consensus Engine and the Validator will be.
+The names, parameters, and exact behavior of these methods are subject to
+change based on additional constraints and requirements discovered during
+implementation.
+
+The methods defined by the Consensus Engine interface are split into the
+following groups:
+
+1. P2P Messaging
+2. Event Handlers
+3. Block Creation
+4. Block Management
+5. Queries
+6. Engine
+
+### P2P Messaging Methods
+The following methods are provided to Consensus Engines for sending consensus
+messages to other nodes on the network. These methods support R1.
+
+| Method | Description |
+| --- | --- |
+| SendTo(peer, message) | Send a consensus message to a specific, connected peer |
+| Broadcast(message) | Broadcast a message to all peers |
+
+### Event Handlers
+The following methods will be implemented by Consensus Engines. They are used
+to update the Consensus Engine on external events. These methods support R1-R3.
+
+| Method | Description |
+| --- | --- |
+| OnMessageReceived(message) | Called when a new consensus message is received |
+| OnNewBlockReceived(block) | Called when a new block is received and validated |
+| OnAddPeer(peer) | Called when a new peer is added |
+| OnDropPeer(peer) | Called when a peer is dropped |
+
+### Block Creation Methods
+The following methods will be provided to Consensus Engines for controlling
+block creation. These methods support R3.
+
+| Method | Description |
+| --- | --- |
+| InitializeBlock() |  Initialize a new block based on the current chain head and start adding batches to it. |
+| FinalizeBlock(data) -> block_id | Stop adding batches to the current block and finalize it. Include the given consensus data in the block. If this call is successful, OnNewBlockReceived() will be called with the block. |
+| CancelBlock() |  Stop adding batches to the current block and abandon it. |
+
+### Block Management Methods
+The following methods will be provided to Consensus Engines for controlling
+chain updates. These methods support R3.
+
+| Method | Description |
+| --- | --- |
+| CommitBlock(block_id) | Set the chain head to the given block id. The Chain Controller would handle the details here. |
+| DropBlock(block_id) | Remove the given block from the system. The main purpose of this method is to allow the Consensus Engine to notify the rest of the system that a block is invalid from the perspective of consensus. |
+
+### Query Methods
+The following methods will be provided to Consensus Engines for getting
+information from the validator.
+
+| Method | Description |
+| --- | --- |
+| GetSetting(setting) -> data | Read the current value of the setting. Supports R4. |
```

Comment:
Is this a convenience method wrapping GetState(address) or does it imply settings outside of global state?